Add status change event

evidolob opened the pull request:
## Solution/Idea
PR adds event API and new SSE(server sent event) channel `status_change`.

## Proposed changes
PR adds event API implementation in `pkg/crc/api/events/events.go`
Modify `sync` client implementation to trigger event when status is changed.
Fixes in existing log events, rename constants related to SSE

## Testing

1. launch demon
2. connect to `status_change` channel with `curl -GET -i --unix-socket <path to socket>/crc-http.sock "http://localhost/events?stream=status_change"`
3.  start crc with `curl -GET -i --unix-socket ~/.crc/crc-http.sock  "http://localhost/api/start"`
4. Ensure that new events appeared, like:
```
id:                                                                                                                            │
data: {"status":{"CrcStatus":"Starting","OpenshiftStatus":"Starting","OpenshiftVersion":"","PodmanVersion":"","DiskUse":0,"DiskSize":0,"RAMUse":0,"RAMSize":0,"Preset":"microshift"}}
event: status_change
```

evidolob commented:
/retest

anjannath commented:
> 1. connect to status_change channel with curl -GET -i --unix-socket <path to socket>/crc-http.sock "http://localhost/events?stream=status_change"
> 2. start crc with crc start

the events are shown only when starting via the api `curl --unix-socket ~/.crc/crc-http.sock "http://localhost/api/start"` running the `crc start` command the events are not shown, is this expected?

anjannath commented on pkg/events/emitter.go:
```diff
@@ -0,0 +1,54 @@
+
+func (e *event[T]) Fire(event T) {
+	e.eventMutex.Lock()
+	defer e.eventMutex.Unlock()
```
is the locking here needed, we are only accessing the map and not adding/removing elements from it

anjannath commented on pkg/crc/api/events/log_stream.go:
```diff
@@ -56,7 +56,12 @@ func (s *streamHook) Fire(entry *logrus.Entry) error {
+	// remove "Line Feed"("\n") character which add was added by json.Encoder
+	if line[len(line)-1] == 10 {
+		line = line[:len(line)-1]
+	}
```
you could also use `bytes.TrimSuffix` here, https://pkg.go.dev/bytes#TrimSuffix

evidolob commented:
> the events are shown only when starting via the api `curl --unix-socket ~/.crc/crc-http.sock "http://localhost/api/start"` running the `crc start` command the events are not shown, is this expected?

Yes, that is expected, as `start`, `stop` is performed on CLI side, not on daemon, at least for now.

anjannath commented:
Tested locally and functionality is working as expected, LGTM. it'd be nice to have more detailed commit messages explaining the changes in each commit

cfergeau commented on cmd/crc/cmd/status.go:
```diff
@@ -81,7 +81,7 @@ func runWatchStatus(writer io.Writer, client *daemonclient.Client, cacheDir stri
 	isPullInit := false
 
-	err := client.SSEClient.Status(func(loadResult *types.ClusterLoadResult) {
+	err := client.SSEClient.ClusterLoad(func(loadResult *types.ClusterLoadResult) {
```
Not obvious why the renaming is needed? The commit log should explain this so that we all have the same understanding of the need for it.

gbraad replied:
It is not actually a rename, but rather a 'prepare'. The way this will work is to enable the status to be sent when requested (starts listening).

Yevhen can explain, as I asked him the same on one of our meetings.

evidolob replied:
Yes, I rename that to avoid confusion with name `Status` vs `StatusChange`

cfergeau commented on pkg/events/emitter.go:
```diff
@@ -0,0 +1,54 @@
+	Fire(data T)
+}
+
+type Disposable func()
```
Dunno if this should use `io.Closer()` https://pkg.go.dev/io#Closer ?

evidolob replied:
We could, but would add unnecessarily complex to code, as `io.Closer` is an interface not a function as `Disposable`

cfergeau commented on pkg/events/emitter.go:
```diff
@@ -0,0 +1,54 @@
+}
+
+func (e *event[T]) AddListener(listener Listener[T]) Disposable {
+	e.eventMutex.Lock()
```
`defer e.eventMutex.Unlock()` and remove the explicit one before the `return`?

cfergeau commented on pkg/events/emitter.go:
```diff
@@ -0,0 +1,54 @@
+	e.eventMutex.Lock()
+	defer e.eventMutex.Unlock()
+	for _, l := range e.listeners {
+		l.Listener(event)
```
`listener.Fire(event)` would read better imo.

cfergeau commented on pkg/events/emitter.go:
```diff
@@ -0,0 +1,54 @@
+	"sync"
+)
+
+type Event[T any] interface {
```
might be worth a small unit test? And a short explanation in the commit log why this is introduced would be useful to have.

cfergeau commented on pkg/crc/machine/status.go:
```diff
@@ -14,6 +14,16 @@ import (
+			CrcStatus: state.Novm,
+		}, nil
+	}
+
```
Isn't `no VM` already covered by `if errors.Is(err, errMissingHost(client.name)) {` below? I think you can change it to return `state.NoVM` instead of an error?

cfergeau commented on pkg/crc/machine/state/state.go:
```diff
@@ -10,6 +10,7 @@ const (
 	Stopped  State = "Stopped"
 	Stopping State = "Stopping"
 	Starting State = "Starting"
+	Novm     State = "NoVM"
```
I'd spell it `NoVM`

cfergeau commented on pkg/crc/api/events/log_stream.go:
```diff
@@ -56,7 +58,12 @@ func (s *streamHook) Fire(entry *logrus.Entry) error {
 
-	s.server.Publish(LOGS, &sse.Event{Event: []byte(LOGS), Data: line})
+	// remove "Line Feed"("\n") character which add was added by json.Encoder
+	if line[len(line)-1] == 10 {
```
I think this could be `== '\n'`. But `TrimRight will do nothing if there is no `\n` at the end of the string, so we don't need to test for it.

cfergeau commented on pkg/crc/api/events/event_server.go:
```diff
@@ -54,8 +54,9 @@ func NewEventServer(machine machine.Client) *EventServer {
-	sseServer.CreateStream(LOGS)
-	sseServer.CreateStream(CLUSTER_LOAD)
+	sseServer.CreateStream(Logs)
+	sseServer.CreateStream(ClusterLoad)
```
This belongs in the first commit where you are already doing some renaming.

gbraad replied:
@cfergeau should we group renames and refactors? I do not mind if they happen along the way in a PR... but in this case, you prefer this to happen as early as possible?

cfergeau replied:
In this specific case, the very first commit does 
```
-       sseServer.CreateStream(STATUS)
+       sseServer.CreateStream(CLUSTER_LOAD)
```
and then in the last commit we have
```
diff --git a/pkg/crc/api/events/event_server.go b/pkg/crc/api/events/event_server.go
index 80ba0dc52..1a7e5155c 100644
--- a/pkg/crc/api/events/event_server.go
+++ b/pkg/crc/api/events/event_server.go
@@ -54,8 +54,9 @@ func NewEventServer(machine machine.Client) *EventServer {
                stream.RemoveSubscriber(sub)
        }
 
-       sseServer.CreateStream(LOGS)
-       sseServer.CreateStream(CLUSTER_LOAD)
+       sseServer.CreateStream(Logs)
+       sseServer.CreateStream(ClusterLoad)
+       sseServer.CreateStream(StatusChange)
        return eventServer
 }
``` 
Feels a bit pointless to change the same line twice in the same PR. The last commit being the tricky one, it's better for reviewers to make it as small as possible (ie avoid unrelated distractions), so I'd move the renames out of it in this case. First commit could be "Change LOGS/STATUS to Logs/Status" followed by "Rename Status to ClusterLoad", this way we don't ever see "CLUSTER_LOAD".

evidolob replied:
I rewrite the commits, hopefully now it looks better

cfergeau commented on pkg/crc/api/events/status_change_stream.go:
```diff
@@ -0,0 +1,62 @@
+package events
+
+import (
+	"encoding/json"
```
Can the commit log contain some background information explaining why we are doing this change, what is our goal, ... ?

cfergeau reviewed (commented):
Need to spend more time on the last commit, but here are some initial comments.

evidolob committed:
Rename 'status' to 'cluster_load' event stream.

This allows to avoid confusion with upcoming status change event stream.

Signed-off-by: Yevhen Vydolob <yvydolob@redhat.com>

evidolob committed:
Add events implementation

It allows us to have unified way to propagate state changes across an app

Co-Authored-By: Christophe Fergeau <cfergeau@redhat.com>
Signed-off-by: Yevhen Vydolob <yvydolob@redhat.com>

evidolob committed:
Add "NoVM" state

This would simplify podman-desktop ext status handling

Signed-off-by: Yevhen Vydolob <yvydolob@redhat.com>

evidolob committed:
Add status change event

Will propagated to client via SSE in ith own channel, to inform clients about CRC status change.
Status modified in "status_change_stream.go", as current status implementation has no single point of change, so tracking transition between states is challenging.

Signed-off-by: Yevhen Vydolob <yvydolob@redhat.com>

evidolob committed:
Fix logging channel messages.

Having '\n' character in between message field is braking SSE specification and client could not parse correctly such events

Signed-off-by: Yevhen Vydolob <yvydolob@redhat.com>

cfergeau commented on pkg/crc/api/events/cluster_load_stream.go:
```diff
@@ -19,18 +19,18 @@ type TickListener struct {
 }
 
-func NewStatusListener(machine crcMachine.Client) EventProducer {
+func newStatusListener(machine crcMachine.Client) EventProducer {
```
This one was missed in the rename

cfergeau commented on pkg/crc/api/events/log_stream.go:
```diff
@@ -23,7 +25,7 @@ func newSSEStreamHook(server *sse.Server) *streamHook {
 			TimestampFormat:   "",
 			DisableTimestamp:  false,
-			DisableHTMLEscape: false,
+			DisableHTMLEscape: true,
```
Can it cause parsing issues at the other hand if HTML is not escaped?

evidolob replied:
Yes, especially new line characters

cfergeau replied:
I don't fully understand what this solves, we remove the new line characters with `TrimRight`, so `DisableHTMLEscape: true` must be solving a different problem?

evidolob replied:
True, delete this change

cfergeau commented on pkg/crc/api/events/status_change_stream.go:
```diff
@@ -0,0 +1,62 @@
+}
+
+func newStatusChangeListener(client machine.Client) EventProducer {
+	return &statusChange{
```
This type could be named `statusChangeListener`

cfergeau commented on pkg/crc/api/events/status_change_stream.go:
```diff
@@ -0,0 +1,62 @@
+			}
+
+		}
+		data, _ := json.Marshal(event)
```
I don't think we should ignore this error

cfergeau commented on pkg/events/events.go:
```diff
@@ -0,0 +1,14 @@
+}
+
+var (
+	StatusChanged = NewEvent[StatusChangedEvent]()
```
Not sure about the global variable here, but we can refine this later.

cfergeau commented on pkg/events/emitter.go:
```diff
@@ -0,0 +1,54 @@
+	}
+}
+
+func (e *event[T]) AddListener(listener Listener[T]) Disposable {
```
For some reason, I expected `Listener[T]` to be a struct or interface, while it's actually a callback func. In my opinion it's clearer to name it `ListenerFunc[T]`.
However, I think we can use an interface here instead of a callback func, this removes the need for `keyCounter`, for `Disposable`, for `listenerWrapper`, ... and in my opinion makes the code easier to read and follow. I've done that in https://github.com/cfergeau/crc/commit/24930c029e46f6dc95172decb033bc85319da972

cfergeau replied:
(I've added a 2nd commit to fix compilation of the testcase)

cfergeau commented on pkg/events/emitter.go:
```diff
@@ -0,0 +1,54 @@
+}
+
+func (e *event[T]) Fire(event T) {
+	for _, listener := range e.listeners {
```
This is missing `e.eventMutex.Lock()` and `defer e.eventMutex.Unlock()`

cfergeau commented on pkg/crc/api/events/status_change_stream.go:
```diff
@@ -0,0 +1,62 @@
+				CrcStatus: state.Error,
+			}, Error: err.Error()}
+		} else {
+			status.CrcStatus = changedEvent.State // override with actual reported state
```
why is the cluster state not matching the reported state? This is the key thing to explain in the comment imo.

cfergeau reviewed (commented):
Another round of comments, still need to spend more time on the pkg/crc/machine/sync.go changes.

cfergeau commented on pkg/crc/machine/sync.go:
```diff
@@ -92,6 +98,13 @@ func (s *Synchronized) Start(ctx context.Context, startConfig types.StartConfig)
+		events.StatusChanged.Fire(events.StatusChangedEvent{State: startResult.Status})
+	} else {
+		events.StatusChanged.Fire(events.StatusChangedEvent{State: state.Error, Error: err})
+	}
```
(commenting here, but that applies to most of the file).
Duplicating the state logic in this file ("after calling `Start()`, the VM is in the `Start` state, unless there was an error, in which case it's in the `Error` state") will likely cause additional maintainance burden, as we have to make sure to keep in sync the state logic in this file, and the state logic in other parts of crc - the status events may be reporting a different state from `Status()`.

Can we emit a `StatusChangedEvent` to tell other code the status changed, but not specify the current status in the event payload? This way the receiver of the event can check `Status()` if they need the status, and we don't have 2 source for the status to keep in sync?

Regarding errors, it seems if `Start()` (or other commands) return an error, you consider that the cluster is in an error state? This makes sense for some commands, but I don't think this matches what is reported by `Status()` at the moment? Would be nice to fix the reported Status() when there are start failures for examples, but doing it here would be papering over the real issue in my opinion.
And in some other cases, errors are less bad, for example in the power off case, maybe the command failed early, and the cluster is still nicely running?

evidolob replied:
>Duplicating the state logic in this file ("after calling Start(), the VM is in the Start state, unless there was an error, in which case it's in the Error state") will likely cause additional maintainance burden, as we have to make sure to keep in sync the state logic in this file, and the state logic in other parts of crc - the status events may be reporting a different state from Status().

Not sure to understand how to resolve you concern, the problem is that we don't have a way to track state change(at least I wasn’t able to find it, it would be nice if you point me that place if if it exist)


> Can we emit a StatusChangedEvent to tell other code the status changed, but not specify the current status in the event payload? This way the receiver of the event can check Status() if they need the status, and we don't have 2 source for the status to keep in sync?

I first do that, but I face a race condition with updating state, for example:
User call start, we fire state change event, before actuality call `Start`, and if listener is fast enough, it could get old state (basically we could lost `starting` state in transition from `stoped` to `running` states)

Possible solution to that, may be moving state change event firing from `Synchronized` client in to `client` implementation of `machine.Client` interface. And fire events only after actually state is changed.


> Regarding errors, it seems if Start() (or other commands) return an error, you consider that the cluster is in an error state? This makes sense for some commands, but I don't think this matches what is reported by Status() at the moment? Would be nice to fix the reported Status() when there are start failures for examples, but doing it here would be papering over the real issue in my opinion.
And in some other cases, errors are less bad, for example in the power off case, maybe the command failed early, and the cluster is still nicely running?

It is matched, as we send with SSE status reported by `Status()` in addition we also send error message, which come from status change event.

cfergeau replied:
My main concern is https://github.com/crc-org/crc/pull/3903#discussion_r1420717479 , and I'm under the impression that if you have a terminal showing the stream of events, and another running `crc status` every second, then the state between the 2 will not always match, in particular sometimes the event stream will contain `State: state.Error` while `crc status` will not report an error?
Guess I need to run both side by side and compare myself to understand better how this works.

cfergeau replied:
(I understand the current code base is missing a centralized "setState" facility, and that this PR has to find a way of doing this, just trying to understand the limitations in the PR)

evidolob replied:
>My main concern is https://github.com/crc-org/crc/pull/3903#discussion_r1420717479 , and I'm under the impression that if you have a terminal showing the stream of events, and another running crc status every second, then the state between the 2 will not always match, in particular sometimes the event stream will contain State: state.Error while crc status will not report an error?
Guess I need to run both side by side and compare myself to understand better how this works.

No, output should be exactly the same, to get that I add
 https://github.com/crc-org/crc/pull/3903/files#diff-e5710ade03da511d4845f380cd63319bf7329e960c1afbbba7fbed4060fcf8edR37 
and 
https://github.com/crc-org/crc/pull/3903/files#diff-e5710ade03da511d4845f380cd63319bf7329e960c1afbbba7fbed4060fcf8edR44-R50

Otherwise we need to move event triggering deeper in crc code, and fire event after execution of pice of code which change `crc status` output.

And it is a question for all, should we integrate events in more depths of CRC core(which leads to spreading events over our codebase) or keep as is( in relative hi level Client interface implementation )?

@gbraad @cfergeau @praveenkumar WDYT?

cfergeau replied:
> And it is a question for all, should we integrate events in more depths of CRC core(which leads to spreading events over our codebase) or keep as is( in relative hi level Client interface implementation )?

I'd say we can go with this for now, more details about the design choices, the implementation limitations, ...

One thing we can explore over time some centralized tracking of the cluster state, a `StateTracker` type, we'd add some calls to `StateTracker.SetState(...)` in the appropriate places, the `StateTracker` would fire events on status changes, it could have a `CanChangeState(state)` helper if needed to know if calling `Stop` will result in an error or not, ...
As long as all `crc` commands do not go through the daemon, it could do some polling to reconcile the internal state with changes through external `crc` calls, ...

cfergeau commented on pkg/events/emitter.go:
```diff
@@ -0,0 +1,46 @@
+	e.eventMutex.Lock()
+	defer e.eventMutex.Unlock()
+	for _, listener := range e.listeners {
+		go listener.Notify(event)
```
I wonder if this is affected by the bug decribed here? https://go.dev/doc/faq#closures_and_goroutines

evidolob replied:
Yes, it is. Something knew learned about golang. 
Fixed.

praveenkumar commented:
/need-rebase

cfergeau commented:
I've rebased it at https://github.com/cfergeau/crc/tree/stus-change-event and added a fixup commit to fix compliation of the first commit in the series.

cfergeau commented:
I'm not sure step 3. is correct in the testing instructions:
> 3. start crc with crc start

Don't we need to start it with the REST API for now as otherwise `start` won't get through the daemon?
```
 curl -GET -i --unix-socket ~/.crc/crc-http.sock  "http://localhost/api/start"
```

I was wondering if `curl -GET -i --unix-socket ~/.crc/crc-http.sock  "http://localhost/api/start"` should also send a status event to help with syncing state if the state change behind the daemon's back?

evidolob commented:
> Don't we need to start it with the REST API for now as otherwise `start` won't get through the daemon?

Yes, you are correct, need to start with rest API.

> 
> ```
>  curl -GET -i --unix-socket ~/.crc/crc-http.sock  "http://localhost/api/start"
> ```
> 
> I was wondering if `curl -GET -i --unix-socket ~/.crc/crc-http.sock "http://localhost/api/start"` should also send a status event to help with syncing state if the state change behind the daemon's back?

Daemon send response only after start is finished(or error incase of any error), but calling `api/start` **will** trigger events on `events?stream=status_change` event channel.

cfergeau commented:
Few more things I noticed during testing, if VM exists, and then I call `delete`, I get
```
id: 
data: {"status":{"CrcStatus":"Error","OpenshiftStatus":"","OpenshiftVersion":"","PodmanVersion":"","DiskUse":0,"DiskSize":0,"RAMUse":0,"RAMSize":0,"Preset":""},"error":"Cannot load 'crc' virtual machine: no such libmachine vm: crc"}
event: status_change
```
which does not seem correct as the VM existed before being deleted.
After that, if I call again `delete` I don't get any notifications. If I call `stop`, I get first `stopping` and then `error`

Probably something that can improve with the `NoVM` state.

evidolob commented:
> Few more things I noticed during testing, if VM exists, and then I call `delete`, I get
> 
> ```
> id: 
> data: {"status":{"CrcStatus":"Error","OpenshiftStatus":"","OpenshiftVersion":"","PodmanVersion":"","DiskUse":0,"DiskSize":0,"RAMUse":0,"RAMSize":0,"Preset":""},"error":"Cannot load 'crc' virtual machine: no such libmachine vm: crc"}
> event: status_change
> ```
> 
> which does not seem correct as the VM existed before being deleted. After that, if I call again `delete` I don't get any notifications. If I call `stop`, I get first `stopping` and then `error`

That was actually bug, `status` was not report `NoVm` state, only error. (It was due to not implementing `Is` method for `MissingHostError`, so comparing error was always wrong)

Now it should work properly.

cfergeau commented on pkg/crc/api/events/status_change_stream.go:
```diff
@@ -0,0 +1,69 @@
+
+func (st *statusChangeListener) Notify(changedEvent events.StatusChangedEvent) {
+	logging.Debugf("State Changed Event %s", changedEvent)
+	var event statusChangeEvent
```
Took me a while to realize there's a `statusChangeEvent` type and a `events.StatusChangedEvent` type, and that both are different, I could not understand how we serialized `changedEvent.State` to something with `Status` in the json :-/ 
Maybe naming can be improved? `changedEvent events.StateChangedEvent`? Still too close to the other name, but at least there's a small difference. Maybe we can come up with something even better.

evidolob replied:
@cfergeau I change type name, hopefully it is more clear now

cfergeau commented on pkg/crc/api/events/status_change_stream.go:
```diff
@@ -0,0 +1,69 @@
+	} else {
+		// event could be fired, before actual code, which change state is called
+		// so status could contain 'old' state, replace it with state received in event
+		status.CrcStatus = changedEvent.State // override with actual reported state
```
This is the part I find questionable in this PR. Instead of having a single source of truth for the status, we now have 2, some of the status comes from `st.machineClient.Status()`, but some of the status is hardcoded in `machine/sync.go`. If the 2 get out of sync, or if we want to improve status reporting in general, we'll have multiple places to change.

cfergeau replied:
Spending more time on it, I understand more the problems this is trying to solve.
One example is when the VM is stopped. `crc status` will succeed and report that crc is stopped, however `crc stop` will return an error and say the machine is already stopped.

Without `status.CrcStatus = changedEvent.State`, when invoking `api/stop`, we would not be reporting an error in the stream, but instead we'd only say the machine is stopped. This would not match what was reported to the `api/stop` call. There are most likely other state transitions where this is needed.

The comment however seems to indicate sometimes `Status` could report a stale status, I haven't seen that in the little testing I did, but I'd be interested in understanding more when this happens? If we don't record this knowledge now when it's more or less fresh in our minds, it will be very difficult to find this back in 6 months from now if we need it :-/

cfergeau replied:
ah, there's more details in an older discussion:
> User call start, we fire state change event, before actuality call Start, and if listener is fast enough, it could get old state (basically we could lost starting state in transition from stoped to running states)

cfergeau replied:
In my opinion, this is an indication that these events are not fired in the right place, and this is papered over with even more complexity :-/

gbraad replied:
> t instead we'd only say the machine is stopped. This would not match what was reported to the api/stop

The API matches the usecase. Stop should in my opinion not return an error when already stopped; as you succeed in doing what is requested. What more information do we actually give with saying this failed as it was already stopped?!

If you believe this should be fixed or recorded, make these states clear for the convergence.

cfergeau replied:
The part you quoted is a description of the behaviour in this PR. These are some notes I took during review/testing when I was trying to understand the design choices which led to the code being reviewed.

evidolob replied:
I try to replace this code, with firing event at more "proper" place, but face problem that `starting` state is "artificial", ie exist only in `sync.go` file, and VM doesn't have/reports `starting` state. Also `stopping` state is the same. 
So I have no other ideas, than accept current PR and think on better state handling, like @cfergeau propose  there https://github.com/crc-org/crc/pull/3903#discussion_r1476135076

Or change state handling first, and then rebase this PR on top of it.

@gbraad @cfergeau WDYT?

cfergeau replied:
I think we can go forward with the current code, but we really need more details in the commit log, so that it records the design choices you made when writing the code, so that we know the current limitations, why this overwrites the state, ...

gbraad commented:
What is holding back this PR?

cfergeau commented:
> What is holding back this PR?

Some discussion around https://github.com/crc-org/crc/pull/3903#discussion_r1476130780 would have been useful

Apart from this, as said multiple times, the comit log must be a lot more detailed about how this works:
>  If we don't record this knowledge now when it's more or less fresh in our minds, it will be very difficult to find this back in 6 months from now if we need it :-/

> I'd say we can go with this for now, more details about the design choices, the implementation limitations, ... 
(this sentence is malformed though :-/ )

And https://github.com/crc-org/crc/pull/3903#discussion_r1474772073 is also still pending

evidolob commented:
@cfergeau Could you look on this?

evidolob commented:
Closing this, as priority changed. I keep the branch in case if we want to return to this.
